### PR TITLE
test: address self-announcement

### PR DESCRIPTION
**Pull request by 0xB10C**
Test that a node sends a self-announcement with its external IP to in- and outbound peers after connection open and again sometime later.

Since the code for the test is mostly the same for addr and addrv2 messages, I opted to add a new test file instead of having duplicate code in `p2p_addr_relay.py` and `p2p_addrv2_relay.py`.

### CODE REVIEW

File: test/functional/p2p_addr_selfannouncement.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        self.log.info(f"Check that we get more self-announcements sometime later (inbound, {addr_version})")
+        for _ in range(5):
+            last_self_announcements_received = addr_receiver.self_announcements_received
+            last_addr_messages_received = addr_receiver.addr_messages_received
+            last_addresses_received = addr_receiver.addresses_received
+            with self.nodes[0].assert_debug_log([f'Advertising address {IP_TO_ANNOUNCE}:{port}']):
+                # m_next_local_addr_send and AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL:
+                # self-announcements are sent on an exponential distribution with mean interval of 24h.
+                # Setting the mocktime 20d forward gives a probability of (1 - e^-(480/24)) that
+                # the event will occur (i.e. this fails once in ~500 million repeats).
```
Comment (0xB10C) [inline test/functional/p2p_addr_selfannouncement.py]: inspired by https://github.com/bitcoin/bitcoin/blob/cca113f5b022b131a78579ab233d0d40643c7c2b/test/functional/p2p_addr_relay.py#L130-L133 (using the same exponent of e^-20 here)

File: test/functional/p2p_addr_selfannouncement.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                addr_receiver.sync_with_ping()
+
+            assert_equal(addr_receiver.self_announcements_received, last_self_announcements_received + 1)
+            assert_equal(addr_receiver.addr_messages_received, last_addr_messages_received + 1)
+            assert_equal(addr_receiver.addresses_received, last_addresses_received + 1)
+
+        self.nodes[0].disconnect_p2ps()
+        # reset mocktime
+        self.mocktime = int(time.time())
+        self.nodes[0].setmocktime(0)
```
Comment (maflcko) [inline test/functional/p2p_addr_selfannouncement.py]: nit: The whole test relies on mocktime, so is there any value in resetting it between the test cases? Also you can use `bumpmocktime` to avoid the manual `+=` + `setmocktime`.

File: test/functional/p2p_addr_selfannouncement.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+class SelfAnnouncementReceiver(P2PInterface):
+    self_announcements_received = 0
+    addresses_received = 0
+    addr_messages_received = 0
+
+    expected = None
+    addrv2_test = False
+
+    def __init__(self, expected, addrv2):
```
Comment (maflcko) [inline test/functional/p2p_addr_selfannouncement.py]:
```suggestion
    def __init__(self, *, expected, addrv2):
```

nit: To force named args, which you are already using.

File: test/functional/p2p_addr_selfannouncement.py (hunk 2 above)
Comment (0xB10C) [reply test/functional/p2p_addr_selfannouncement.py]: Done! Wasn't aware of `bumpmocktime`, thanks. Cleaned up use of mocktime a bit.

File: test/functional/p2p_addr_selfannouncement.py (hunk 3 above)
Comment (0xB10C) [reply test/functional/p2p_addr_selfannouncement.py]: TIL!
done.

File: test/functional/p2p_addr_selfannouncement.py (hunk 1 above)
Comment (mzumsande) [reply test/functional/p2p_addr_selfannouncement.py]: nothing for this PR, but since these things are quite common (recently #33121), I wonder if a mode `-test=cutoffexponentialtimers` or something similar would make sense which would cut off the long tail of all exponential timers (`rand_exp_duration`) at some percentile, so we don't need to deal with these probabilistic failures in functional tests.

File: test/functional/p2p_addr_selfannouncement.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+class AddrSelfAnnouncementTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [[f"-externalip={IP_TO_ANNOUNCE}"]]
+
+    def run_test(self):
+        # populate addrman to have some addresses for a GETADDR response
+        for i in range(50):
+            first_octet = 1 + (i >> 8)
+            second_octet = i % 256
```
Comment (Bicaru20) [inline test/functional/p2p_addr_selfannouncement.py]: Since increasing the 50 in the loop will not contribute anything to the test, maybe the `>> 8` and  `% 256` could be erased.
```suggestion
            first_octet = 1 + i
            second_octet = i
```

**Comment by Bicaru20** [comment]
Code review ACK fb9c393564ef098267ad2a921b75e0716f38b047. I have run the test and everything checks out. 

 I left a small comment in the code, also fine with merging as-is

File: test/functional/p2p_addr_selfannouncement.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        port = netinfo["localaddresses"][0]["port"]
+        self.nodes[0].setmocktime(int(time.time()))
+
+        expected = CAddress()
+        expected.nServices = int(netinfo["localservices"], 16)
+        expected.ip = IP_TO_ANNOUNCE
+        expected.port = port
+        expected.time = self.nodes[0].mocktime
+
+        self.log.info(f"Check that we get an initial self-announcement on a outbound connection from the node (outbound, {addr_version})")
```
Comment (rkrux) [inline test/functional/p2p_addr_selfannouncement.py]: In fb9c393564ef098267ad2a921b75e0716f38b047 "test: address self-announcement"

Is it intentional to mention outbound twice in this log message and deviate from the similar log message of the inbound test?

File: test/functional/p2p_addr_selfannouncement.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                # the event will occur (i.e. this fails once in ~500 million repeats).
+                self.nodes[0].bumpmocktime(20 * ONE_DAY)
+                addr_receiver.expected.time = self.nodes[0].mocktime
+                addr_receiver.sync_with_ping()
+
+            assert_equal(addr_receiver.self_announcements_received, last_self_announcements_received + 1)
+            assert_equal(addr_receiver.addr_messages_received, last_addr_messages_received + 1)
+            assert_equal(addr_receiver.addresses_received, last_addresses_received + 1)
+
+        self.nodes[0].disconnect_p2ps()
```
Comment (rkrux) [inline test/functional/p2p_addr_selfannouncement.py]: In fb9c393564ef098267ad2a921b75e0716f38b047 "test: address self-announcement"

I noticed some duplication in the test that could be removed with the following diff. I don't feel too strongly about this suggestion and if you think such a de-duplicated code might make it more difficult to update the test in the future, then feel free to not implement.

```diff
diff --git a/test/functional/p2p_addr_selfannouncement.py b/test/functional/p2p_addr_selfannouncement.py
index 65527d1e77..47de6156f8 100755
--- a/test/functional/p2p_addr_selfannouncement.py
+++ b/test/functional/p2p_addr_selfannouncement.py
@@ -69,62 +69,28 @@ class AddrSelfAnnouncementTest(BitcoinTestFramework):
             a = f"{first_octet}.{second_octet}.1.1"
             self.nodes[0].addpeeraddress(a, 8333)
 
-        self.self_announcement_inbound_test(addrv2=False)
-        self.self_announcement_inbound_test(addrv2=True)
-        self.self_announcement_outbound_test(addrv2=False)
-        self.self_announcement_outbound_test(addrv2=True)
-
-    def self_announcement_inbound_test(self, addrv2=False):
-        addr_version = "addrv2" if addrv2 else "addrv1"
-        self.log.info(f"Test that the node does an address self-announcement to inbound connections ({addr_version})")
-
-        # We only self-announce after initial block download is done
-        assert (not self.nodes[0].getblockchaininfo()["initialblockdownload"])
-
-        netinfo = self.nodes[0].getnetworkinfo()
-        port = netinfo["localaddresses"][0]["port"]
-        self.nodes[0].setmocktime(int(time.time()))
-
-        expected = CAddress()
-        expected.nServices = int(netinfo["localservices"], 16)
-        expected.ip = IP_TO_ANNOUNCE
-        expected.port = port
-        expected.time = self.nodes[0].mocktime
-
-        self.log.info(f"Check that we get an initial self-announcement when connecting to a node and sending a GETADDR (inbound, {addr_version})")
-        with self.nodes[0].assert_debug_log([f'Advertising address {IP_TO_ANNOUNCE}:{port}']):
-            addr_receiver = self.nodes[0].add_p2p_connection(SelfAnnouncementReceiver(expected=expected, addrv2=addrv2))
-            addr_receiver.sync_with_ping()
+        self.self_announcement_test(outbound=False, addrv2=False, assert_on_connection_open=self.inbound_connection_open_assertions)
+        self.self_announcement_test(outbound=False, addrv2=True, assert_on_connection_open=self.inbound_connection_open_assertions)
+        self.self_announcement_test(outbound=True, addrv2=False, assert_on_connection_open=self.outbound_connection_open_assertions)
+        self.self_announcement_test(outbound=True, addrv2=True, assert_on_connection_open=self.outbound_connection_open_assertions)
 
+    def inbound_connection_open_assertions(self, addr_receiver):
         # We expect one self-announcement and multiple other addresses in
         # response to a GETADDR in a single addr / addrv2 message.
         assert_equal(addr_receiver.self_announcements_received, 1)
         assert_equal(addr_receiver.addr_messages_received, 1)
         assert_greater_than(addr_receiver.addresses_received, 1)
 
-        self.log.info(f"Check that we get more self-announcements sometime later (inbound, {addr_version})")
-        for _ in range(5):
-            last_self_announcements_received = addr_receiver.self_announcements_received
-            last_addr_messages_received = addr_receiver.addr_messages_received
-            last_addresses_received = addr_receiver.addresses_received
-            with self.nodes[0].assert_debug_log([f'Advertising address {IP_TO_ANNOUNCE}:{port}']):
-                # m_next_local_addr_send and AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL:
-                # self-announcements are sent on an exponential distribution with mean interval of 24h.
-                # Setting the mocktime 20d forward gives a probability of (1 - e^-(480/24)) that
-                # the event will occur (i.e. this fails once in ~500 million repeats).
-                self.nodes[0].bumpmocktime(20 * ONE_DAY)
-                addr_receiver.expected.time = self.nodes[0].mocktime
-                addr_receiver.sync_with_ping()
-
-            assert_equal(addr_receiver.self_announcements_received, last_self_announcements_received + 1)
-            assert_equal(addr_receiver.addr_messages_received, last_addr_messages_received + 1)
-            assert_equal(addr_receiver.addresses_received, last_addresses_received + 1)
-
-        self.nodes[0].disconnect_p2ps()
+    def outbound_connection_open_assertions(self, addr_receiver):
+        # We expect only the self-announcement.
+        assert_equal(addr_receiver.self_announcements_received, 1)
+        assert_equal(addr_receiver.addr_messages_received, 1)
+        assert_equal(addr_receiver.addresses_received, 1)
 
-    def self_announcement_outbound_test(self, addrv2=True):
+    def self_announcement_test(self, outbound, addrv2, assert_on_connection_open):
+        connection_type = "outbound" if outbound else "inbound"
         addr_version = "addrv2" if addrv2 else "addrv1"
-        self.log.info(f"Test that the node does an address self-announcement to outbound connections ({addr_version})")
+        self.log.info(f"Test that the node does an address self-announcement to {connection_type} connections ({addr_version})")
 
         # We only self-announce after initial block download is done
         assert (not self.nodes[0].getblockchaininfo()["initialblockdownload"])
@@ -139,21 +105,19 @@ class AddrSelfAnnouncementTest(BitcoinTestFramework):
         expected.port = port
         expected.time = self.nodes[0].mocktime
 
-        self.log.info(f"Check that we get an initial self-announcement on a outbound connection from the node (outbound, {addr_version})")
+        self.log.info(f"Check that we get an initial self-announcement when connecting to a node and sending a GETADDR ({connection_type}, {addr_version})")
         with self.nodes[0].assert_debug_log([f'Advertising address {IP_TO_ANNOUNCE}:{port}']):
-            addr_receiver = self.nodes[0].add_outbound_p2p_connection(SelfAnnouncementReceiver(expected=expected, addrv2=addrv2), p2p_idx=0, connection_type="outbound-full-relay")
+            addr_receiver = self.nodes[0].add_outbound_p2p_connection(SelfAnnouncementReceiver(expected=expected, addrv2=addrv2), p2p_idx=0, connection_type="outbound-full-relay") if outbound else self.nodes[0].add_p2p_connection(SelfAnnouncementReceiver(expected=expected, addrv2=addrv2))
             addr_receiver.sync_with_ping()
 
-        # We expect only the self-announcement.
-        assert_equal(addr_receiver.self_announcements_received, 1)
-        assert_equal(addr_receiver.addr_messages_received, 1)
-        assert_equal(addr_receiver.addresses_received, 1)
+        assert_on_connection_open(addr_receiver)
 
-        # to avoid the node evicting the outbound peer, protect it by announcing the most recent header to it
-        tip_header = from_hex(CBlockHeader(), self.nodes[0].getblockheader(self.nodes[0].getbestblockhash(), False))
-        addr_receiver.send_and_ping(msg_headers([tip_header]))
+        if outbound:
+            # to avoid the node evicting the outbound peer, protect it by announcing the most recent header to it
+            tip_header = from_hex(CBlockHeader(), self.nodes[0].getblockheader(self.nodes[0].getbestblockhash(), False))
+            addr_receiver.send_and_ping(msg_headers([tip_header]))
 
-        self.log.info(f"Check that we get more self-announcements sometime later (outbound, {addr_version})")
+        self.log.info(f"Check that we get more self-announcements sometime later ({connection_type}, {addr_version})")
         for _ in range(5):
             last_self_announcements_received = addr_receiver.self_announcements_received
             last_addr_messages_received = addr_receiver.addr_messages_received
@@ -173,6 +137,5 @@ class AddrSelfAnnouncementTest(BitcoinTestFramework):
 
         self.nodes[0].disconnect_p2ps()
 
-
 if __name__ == '__main__':
     AddrSelfAnnouncementTest(__file__).main()

```

**Review by rkrux — COMMENTED** [review]
Code review fb9c393564ef098267ad2a921b75e0716f38b047

File: test/functional/p2p_addr_selfannouncement.py (hunk 6 above)
Comment (fjahr) [reply test/functional/p2p_addr_selfannouncement.py]: Not sure if this is the best way to do it but I also had the feeling that some deduplication should be possible here among the inbound and outbound test

**Review by fjahr — COMMENTED** [review]
Concept ACK

**Review by danielabrozzoni — COMMENTED** [review]
tACK fb9c393564ef098267ad2a921b75e0716f38b047

Happy to re-ACK if you decide to pick up https://github.com/bitcoin/bitcoin/pull/34039#discussion_r2610717344 :)

File: test/functional/p2p_addr_selfannouncement.py (hunk 6 above)
Comment (0xB10C) [reply test/functional/p2p_addr_selfannouncement.py]: Thanks rkrux, included!

File: test/functional/p2p_addr_selfannouncement.py (hunk 4 above)
Comment (0xB10C) [reply test/functional/p2p_addr_selfannouncement.py]: done!

**Comment by 0xB10C** [comment]
Included both suggestions, thanks!
- https://github.com/bitcoin/bitcoin/pull/34039#discussion_r2607756886
- https://github.com/bitcoin/bitcoin/pull/34039#discussion_r2610717344

File: test/functional/p2p_addr_selfannouncement.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        self.addrv2_test = addrv2
+
+    def handle_addr_message(self, message):
+        self.addr_messages_received += 1
+        for addr in message.addrs:
+            self.addresses_received += 1
+            if addr == self.expected:
+                self.self_announcements_received += 1
+            # The tricky part of this test is the timestamp of the self-announcement.
+            # The following helped me to debug timestamp mismatches, so I'm leaving it in for the next one.
```
Comment (mzumsande) [inline test/functional/p2p_addr_selfannouncement.py]: I would suggest to rework or remove this before merge, first person language and `print()` don't seem appropriate at that stage anymore, the mismatch print could be an assert if it's not expected to ever happen.

File: test/functional/p2p_addr_selfannouncement.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        self.num_nodes = 1
+        self.extra_args = [[f"-externalip={IP_TO_ANNOUNCE}"]]
+
+    def run_test(self):
+        # populate addrman to have some addresses for a GETADDR response
+        for i in range(50):
+            a = f"{1 + i}.{i}.1.1"
+            self.nodes[0].addpeeraddress(a, 8333)
+
+        self.self_announcement_test(outbound=False, addrv2=False, assert_on_connection_open=self.inbound_connection_open_assertions)
```
Comment (mzumsande) [inline test/functional/p2p_addr_selfannouncement.py]: `assert_on_connection_open` seems like a redundant arg, could call one or the other based on the existing `outbound` arg.

File: test/functional/p2p_addr_selfannouncement.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+        expected = CAddress()
+        expected.nServices = int(netinfo["localservices"], 16)
+        expected.ip = IP_TO_ANNOUNCE
+        expected.port = port
+        expected.time = self.nodes[0].mocktime
+
+        with self.nodes[0].assert_debug_log([f'Advertising address {IP_TO_ANNOUNCE}:{port}']):
+            if outbound:
+                self.log.info(f"Check that we get an initial self-announcement on a outbound connection from the node ({connection_type}, {addr_version})")
```
Comment (mzumsande) [inline test/functional/p2p_addr_selfannouncement.py]: nit: **an** outbound

File: test/functional/p2p_addr_selfannouncement.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        self.self_announcement_test(outbound=False, addrv2=True, assert_on_connection_open=self.inbound_connection_open_assertions)
+        self.self_announcement_test(outbound=True, addrv2=False, assert_on_connection_open=self.outbound_connection_open_assertions)
+        self.self_announcement_test(outbound=True, addrv2=True, assert_on_connection_open=self.outbound_connection_open_assertions)
+
+    def inbound_connection_open_assertions(self, addr_receiver):
+        # We expect one self-announcement and multiple other addresses in
+        # response to a GETADDR in a single addr / addrv2 message.
+        assert_equal(addr_receiver.self_announcements_received, 1)
+        assert_equal(addr_receiver.addr_messages_received, 1)
+        assert_greater_than(addr_receiver.addresses_received, 1)
```
Comment (mzumsande) [inline test/functional/p2p_addr_selfannouncement.py:76]: note: I first thought that there might be a race condition (node receives verack first, and getaddr with some delay, sends out self-advertisment in between) but this isn't actually true because in this case we won't self-advertise due to #21528 (`SetupAddressRelay` not called yet) . So this is fine for now (whether the node behavior of mixing these is good is another issue)

**Review by mzumsande — COMMENTED** [review]
Concept ACK

File: test/functional/p2p_addr_selfannouncement.py (hunk 8 above)
Comment (0xB10C) [reply test/functional/p2p_addr_selfannouncement.py]: done.

File: test/functional/p2p_addr_selfannouncement.py (hunk 7 above)
Comment (0xB10C) [reply test/functional/p2p_addr_selfannouncement.py]: Dropped it.

**Comment by 0xB10C** [comment]
Addressed comments by mzumsande

**Review by rkrux — APPROVED** [review]
ACK 1841bf9

Thanks for the co-author attribution.

**Comment by Bicaru20** [comment]
ACK https://github.com/bitcoin/bitcoin/commit/1841bf9cb67bca99273b3efa38633113422f5882

File: test/functional/p2p_addr_selfannouncement.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+in- and outbound peers after connection open and again sometime later.
+"""
+
+import time
+
+from test_framework.messages import (
+    CAddress,
+    from_hex,
+    msg_headers,
+    CBlockHeader,
```
Comment (fjahr) [inline test/functional/p2p_addr_selfannouncement.py:16]: nit: Should be below `CAddress`

File: test/functional/p2p_addr_selfannouncement.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        assert_greater_than(addr_receiver.addresses_received, 1)
+
+    def outbound_connection_open_assertions(self, addr_receiver):
+        # We expect only the self-announcement.
+        assert_equal(addr_receiver.self_announcements_received, 1)
+        assert_equal(addr_receiver.addr_messages_received, 1)
+        assert_equal(addr_receiver.addresses_received, 1)
+
+    def self_announcement_test(self, *, outbound, addrv2):
+        connection_type = "outbound" if outbound else "inbound"
```
Comment (fjahr) [inline test/functional/p2p_addr_selfannouncement.py:85]: nit: Could have passed the strings as arguments directly instead of interpreting the bools, this would make the call-site a bit easier to read as well I think.

File: test/functional/p2p_addr_selfannouncement.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+        addr_version = "addrv2" if addrv2 else "addrv1"
+        self.log.info(f"Test that the node does an address self-announcement to {connection_type} connections ({addr_version})")
+
+        # We only self-announce after initial block download is done
+        assert (not self.nodes[0].getblockchaininfo()["initialblockdownload"])
```
Comment (fjahr) [inline test/functional/p2p_addr_selfannouncement.py:90]: Is this really needed? Since we are loading the preconfigured chain here I don't think we could ever hit this. I guess this is here rather for documentation purposes? Even if it is a problem then this could prbably move to `run_test` to just run once.

File: test/functional/p2p_addr_selfannouncement.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+            with self.nodes[0].assert_debug_log([f'Advertising address {IP_TO_ANNOUNCE}:{port}']):
+                # m_next_local_addr_send and AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL:
+                # self-announcements are sent on an exponential distribution with mean interval of 24h.
+                # Setting the mocktime 20d forward gives a probability of (1 - e^-(480/24)) that
+                # the event will occur (i.e. this fails once in ~500 million repeats).
+                self.nodes[0].bumpmocktime(20 * ONE_DAY)
+                addr_receiver.expected.time = self.nodes[0].mocktime
+                addr_receiver.sync_with_ping()
+
+            assert_equal(addr_receiver.self_announcements_received, last_self_announcements_received + 1)
```
Comment (fjahr) [inline test/functional/p2p_addr_selfannouncement.py:135]: nit: This mostly repeats the checks from `inbound_connection_open_assertions` and its outbound counterpart with different numbers. I think there could be some further de-duplication where you replace it with a more generic method on the addr_receiver just called `connection_open_assertions` and then give it the number as args. But it's not a big deal.

File: test/functional/p2p_addr_selfannouncement.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+        else:
+            self.inbound_connection_open_assertions(addr_receiver)
+
+        if outbound:
+            # to avoid the node evicting the outbound peer, protect it by announcing the most recent header to it
+            tip_header = from_hex(CBlockHeader(), self.nodes[0].getblockheader(self.nodes[0].getbestblockhash(), False))
+            addr_receiver.send_and_ping(msg_headers([tip_header]))
+
+        self.log.info(f"Check that we get more self-announcements sometime later ({connection_type}, {addr_version})")
+        for _ in range(5):
```
Comment (fjahr) [inline test/functional/p2p_addr_selfannouncement.py:122]: Just curious: Why do this 5 times? Is there any special significance to it? Based on the log above I would have expected the loop to exit once a first self-announcement has been observed.

Maybe the info log could be reworded a bit to "Check that we get repeated self-announcements periodically".

File: test/functional/p2p_addr_selfannouncement.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+                self.log.info(f"Check that we get an initial self-announcement when connecting to a node and sending a GETADDR ({connection_type}, {addr_version})")
+                addr_receiver = self.nodes[0].add_p2p_connection(SelfAnnouncementReceiver(expected=expected, addrv2=addrv2))
+            addr_receiver.sync_with_ping()
+
+        if outbound:
+            self.outbound_connection_open_assertions(addr_receiver)
+        else:
+            self.inbound_connection_open_assertions(addr_receiver)
+
+        if outbound:
```
Comment (fjahr) [inline test/functional/p2p_addr_selfannouncement.py:116]: nit: Could move the code block in this if into the above outbound conditional.

**Review by fjahr — COMMENTED** [review]
Code review ACK 1841bf9cb67bca99273b3efa38633113422f5882

Left a bunch of nits and minor questions. But I think this is also fine to merge as is.

**Comment by achow101** [comment]
ACK 1841bf9cb67bca99273b3efa38633113422f5882
